tools/donate_cpu_lib.py

```diff
@@ -35,42 +37,58 @@ def check_requirements():
-        if os.path.exists(cppcheck_path):
-            try:
-                os.chdir(cppcheck_path)
-                try:
-                    subprocess.check_call(['git', 'checkout', '-f', 'main'])
-                except subprocess.CalledProcessError:
-                    subprocess.check_call(['git', 'checkout', '-f', 'master'])
-                    subprocess.check_call(['git', 'pull'])
-                    subprocess.check_call(['git', 'checkout', 'origin/main', '-b', 'main'])
-                subprocess.check_call(['git', 'pull'])
-            except:
-                print('Failed to update Cppcheck sources! Retrying..')
-                time.sleep(10)
-                continue
-        else:
-            try:
-                subprocess.check_call(['git', 'clone', 'https://github.com/danmar/cppcheck.git', cppcheck_path])
-            except:
-                print('Failed to clone, will try again in 10 minutes..')
-                time.sleep(600)
-                continue
-        time.sleep(2)
-        return True
-    if os.path.exists(cppcheck_path):
-        print('Failed to update Cppcheck sources, trying a fresh clone..')
+# Try and retry with exponential backoff if an exception is raised
+def try_retry(fun, fargs=(), max_tries=5):
+    sleep_duration = 5.0
+    for i in range(max_tries):
         try:
-            os.chdir(work_path)
-            shutil.rmtree(cppcheck_path)
-            get_cppcheck(cppcheck_path, work_path)
-        except:
-            print('Failed to remove Cppcheck folder, please manually remove ' + work_path)
-            return False
-    return False
+            return fun(*fargs)
+        except KeyboardInterrupt as e:
+            # Do not retry in case of user abort
+            raise e
+        except BaseException as e:
+            if i < max_tries - 1:
+                print("{} in {}: {}".format(type(e).__name__, fun.__name__, str(e)))
+                print("Trying {} again in {} seconds".format(fun.__name__, sleep_duration))
+                time.sleep(sleep_duration)
+                sleep_duration *= 2.0
+            else:
+                print("Maximum number of tries reached for {}".format(fun.__name__))
+                raise e
+
+
+def clone_cppcheck(repo_path):
+    repo_git_dir = os.path.join(repo_path, '.git')
+    if os.path.exists(repo_git_dir):
+        return
+    # A shallow git clone (depth = 1) is enough for building and scanning.
+    # Do not checkout until fetch_cppcheck_version.
+    subprocess.check_call(['git', 'clone', '--depth=1', '--no-checkout', CPPCHECK_REPO_URL, repo_path])
+    # Checkout an empty branch to allow "git worktree add" for main later on
+    os.chdir(repo_path)
+    try:
+        # git >= 2.27
+        subprocess.check_call(['git', 'switch', '--orphan', 'empty'])
+    except subprocess.CalledProcessError:
+        subprocess.check_call(['git', 'checkout','--orphan', 'empty'])
+
+
+def get_cppcheck_version(repo_path, version, cppcheck_path):
```

Comment:
The function name is misleading. First you expect a `get_*` function to return something. Otherwise it does git stuff which is not obvious from the name.

Reply:
That's fair. What about checkout_cppcheck_version?

Reply:
That's certainly better.